Add data about DMA peripheral triggers

salkinium opened the pull request:
This adds a parser for the `IP/DMA-*` files for all STM32 devices.

- [x] `stm32` data, only has channels (F0, F1, F3, L0, L1, L4)
- [x] `stm32-extended` data, has streams and channels (F2, F4, F7)
- [x] `stm32-advanced` data, what the heck?! (H7, L4+)


cc @rleh @chris-durand

salkinium committed:
[stm32] Parse DMA signal connections

salkinium commented:
There's also some configuration recommendations per trigger:
- `mode` (normal or circular)
- `direction` (m2p or p2m)
- `increment` boolean
- `alignment`
- `burst`

chris-durand commented on tools/generator/dfg/stm32/stm_peripherals.py:
```diff
@@ -105,12 +105,31 @@
+                'hardware': 'stm32-channel-request',
+                'features': [],
+                'protocols': ['mem2mem', 'mem2per', 'per2per'],
+                'devices': [{'family': ['l4']}]
```
STM32F09x devices with the "STM32F091_dma_v1_1" IP also have a request mux like the L4.

salkinium replied:
Yeah, this categorization is still messy, cos I don’t have a good overview of the DMA implementations yet.

chris-durand replied:
I roughly cross-checked the generated device files with the reference manuals. Almost all data looks reasonable. Only the requests for F09x devices are missing.

salkinium replied:
Not just F091 and F098 devices use requests, but also the F030 with size C, and all L0 and L4. ST is weird.

chris-durand commented:
> what the heck?! (H7, L4+)

The H7 has 3 DMA instances (DMA1, DMA2, BDMA) and two DMAMUX.
They are connected in the following way:

- DMAMUX1 out 0-7 -> DMA1 channel 0-7
- DMAMUX1 out 8-15 -> DMA2 channel 8-15
- DMAMUX2 out 0-7 -> BDMA channel 0-7

Any request line input of the mux can be connected to any output. The mapping between request mux input ids and peripherals (Table 116/119 in H7 manual) is specified in the `<RefParameter Comment="Request number" DefaultValue="" Group="" Name="Request" Type="list" Visible="false">` tag in the DMA and BDMA IP xml.

Additionally the DMAMUX also has further trigger and synchronization inputs ...

salkinium commented:
Is there a branch for the DMA driver WIP? I want to see what data could be useful for it.

rleh commented:
The branch 'feature/dmahal' is available on RCAs private Gitlab.

salkinium commented:
Btw, I uploaded some of my experimentation of comparing/compressing the STM32 memory maps as described by the CMSIS header files. Not sure what it all means, but you can get an overview of the different DMA implementations:
- [DMA](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/dma.xml) ([registers only](https://github.com/salkinium/stm32-memory/blob/master/stm32/dma.xml))
- [DMA_Channel](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/dma_channel.xml)
- [DMA_Request](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/dma_request.xml)
- [BDMA](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/bdma.xml)
- [BDMA_Channel](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/bdma_channel.xml)
- [MDMA](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/mdma.xml)
- [MDMA_Channel](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/mdma_channel.xml)
- [DMAMUX_ChannelStatus](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/dmamux_channelstatus.xml)
- [DMAMUX_RequestGenStatus](https://github.com/salkinium/stm32-memory/blob/master/stm32-details/dmamux_requestgenstatus.xml)

rleh commented:
The DMA data is not accessible from the modm build system:
```bash
$ print(device.get_driver("dma"))
{'instance': ['1', '2'], 'name': 'dma', 'type': 'stm32-stream-channel'}
```

salkinium commented:
> The DMA data is not accessible from the modm build system:

What `:target` were you using? Maybe I missed some stm32 family or messed up the device file.

chris-durand commented:
> What :target were you using? Maybe I missed some stm32 family or messed up the device file.

We were using a STM32F4 Discovery example (stm32f407vgt). It seems child elements are discarded  when the parent node has a "value" attribute.

salkinium commented:
> It seems child elements are discarded when the parent node has a "value" attribute.

Oh yeah 🤦‍♂️
I remember now this was an issue with choosing only a subset of XML to later be able to map it losslessly into a py-dict/JSON-like structure. The `value="value"` attribute replaces the `<node>value</node>` to avoid some edge-cases during conversion.
I need to raise an exception when trying to set both value + children.

salkinium commented:
Ok, I fixed this by using the `position` attribute, analog to `core/vector`.
However, in order to preserve the `driver/instance` semantics, I moved the additional information into their separate nodes with an instance attribute `<streams instance="1">`.

rleh commented on devices/stm32/stm32f1-03-c_d_e.xml:
```diff
@@ -141,6 +137,96 @@
     <driver name="usb" type="stm32-v1.1"/>
     <driver name="wwdg" type="stm32-v1.0"/>
+    <driver name="dma" type="stm32-channel">
+      <channels>
```
This `<channels>` does not know to which instance it belongs.

https://github.com/modm-io/modm-devices/blob/d02739c9585e03f84b4693446b0654202dd53623/devices/stm32/stm32f1-03-c_d_e.xml#L227-L228

@salkinium

salkinium replied:
Thanks, fixed!

chris-durand commented:
@salkinium Thanks for the great data.

Unfortunately, some data is still missing. ST wouldn't be ST if they didn't prepare a special surprise for us. Most F0 and all F3 devices have an annoying DMA remapping feature. To use some combinations of peripherals and channels certain bits in SYSCFG_CFGR1 have to be set.

![image](https://user-images.githubusercontent.com/25187160/45596636-0d6dec80-b9bf-11e8-9c9f-ef278f62df3d.png)

salkinium commented:
> ST wouldn't be ST if they didn't prepare a special surprise for us. Most F0 and all F3 devices have an annoying DMA remapping feature. To use some combinations of peripherals and channels certain bits in SYSCFG_CFGR1 have to be set.

Oh, FFS.

salkinium commented:
Can you check if this is somehow encoded in the raw CubeMX data? Otherwise I'd have to manually add it like the AFIO for STM32F1/L1

rleh commented:
There seems to be some data:
E.g. in `db/mcu/IP/DMA-STM32F303_dma_v1_0_Modes.xml`(line 96)
```xml
<RefParameter Comment="DMA Remap" DefaultValue="HAL_REMAPDMA_ADC24_DMA2_CH34" Group="" Name="DMA_Remap" Type="list" Visible="false">
  <PossibleValue Comment="ADC2 and ADC4 DMA remap" Value="HAL_REMAPDMA_ADC24_DMA2_CH34"/>
  <Condition Diagnostic="" Expression="(ADC2_Remap&amp;(Instance=DMA2_Channel3))|(ADC4_Remap&amp;(Instance=DMA2_Channel4))"/>
</RefParameter>
[...]
```
Corresponding table in the F303 datasheet:
![image](https://user-images.githubusercontent.com/2820734/45596815-56737000-b9c2-11e8-85d9-a302c56a7d78.png)

chris-durand commented:
CubeMX generates something like `__HAL_DMA_REMAP_CHANNEL_ENABLE(DMA_REMAP_USART1_TX_DMA_CH4);`
The data we need is in the ST HAL headers, not in CubeMX.

rleh commented:
Something like this:
```xml
<remapping>
  <remap_group>
    <default>
      <signal dma_instance="2" dma_channel="1" driver="adc" instance="2" name="dma"/>
      <signal dma_instance="2" dma_channel="2" driver="adc" instance="4" name="dma"/>
    </default>
    <remapped>
      <signal dma_instance="2" dma_channel="3" driver="adc" instance="2" name="dma"/>
      <signal dma_instance="2" dma_channel="4" driver="adc" instance="4" name="dma"/>
    </remapped>
    <!--
     Remap bit "SYSCFG_CFGR1_ADC24_DMA_RMP" can be computed
     from <signal> elements inside <remapped> above
    -->
  </remap_group>
</remapping>
```

salkinium commented on devices/stm32/stm32f3-03-6_8.xml:
```diff
@@ -133,46 +133,86 @@
+          </signal>
+          <signal driver="tim" instance="17" name="up">
+            <remap position="12" mask="1" id="0"/>
+          </signal>
```
cc @rleh @chris-durand 

I've added these remappings manually from the F303x6/8 RM, so we can discuss the formatting. I believe this is a compact and unique mapping, but I have to check this computationally.

This would require the dma driver module file to compute the conflicts somehow, but the information is all here.

salkinium replied:
Note that similar to the STM32F1 remap, the bit position for the config registers is directly encoded here, to be independent of STs naming register conventions.

salkinium commented on devices/stm32/stm32f3-03-6_8.xml:
```diff
@@ -133,46 +133,86 @@
+          </signal>
+          <signal driver="spi" instance="1" name="rx">
+            <remap position="64" mask="3" id="0"/>
+            <remap position="64" mask="3" id="3"/>
```
`(position / 32) + 1` gives the registers number, here `CFGR3`. Same idea used for AFIOx in STM32F1 GPIO remappings.

salkinium replied:
No clue why ST thinks it needs two mappings for this value. I'm not sure if this isn't a copy-paste error in the PDF.

salkinium committed:
[dfg] Add STM32F0/F3 DMA remap groups

salkinium committed:
[dfg] Rename STM32F1 GPIO remap groups

salkinium commented on tools/generator/dfg/stm32/stm.py:
```diff
@@ -89,6 +89,71 @@ def getGpioRemapForModuleConfig(module, config):
+    'dma1ch2': {
+        'adc2':      {'position': 72, 'clear': 3, 'set': 2},
+        'i2c_tx':    {'position': 70, 'clear': 3, 'set': 1},
+        'spi_rx':    {'position': 64, 'clear': 3, 'set': 0}, # also 'set': 3
```
I don't see an advantage for also adding the 0b11 encoding here, it does the same thing as the 0b00 encoding.

salkinium commented on tools/generator/dfg/stm32/stm.py:
```diff
@@ -89,6 +89,71 @@ def getGpioRemapForModuleConfig(module, config):
+        'i2c_rx':    {'position': 68, 'clear': 3, 'set': 0}, # also 'set': 3
+        'spi_tx':    {'position': 66, 'clear': 3, 'set': 2},
+        'tim17_ch1': 'tim17_up',
+        'tim17_up':  {'position': 12, 'clear': 1, 'set': 1},
```
The STM32 reference manual incredibly inconsistent. All F3 reference manuals say this:
<img width="617" alt="screen shot 2018-09-19 at 19 58 15" src="https://user-images.githubusercontent.com/163066/45772002-ee818b80-bc46-11e8-87cc-7b5f206c0150.png">
With the corresponding table entry:
![screen shot 2018-09-19 at 19 58 48](https://user-images.githubusercontent.com/163066/45772005-f0e3e580-bc46-11e8-872c-0e9242b96d1a.png)

But the F373 reference manual says for this:
![screen shot 2018-09-19 at 19 57 49](https://user-images.githubusercontent.com/163066/45771962-ca25af00-bc46-11e8-902a-22fe83babdce.png)
![screen shot 2018-09-19 at 19 58 08](https://user-images.githubusercontent.com/163066/45771979-dad62500-bc46-11e8-9ef6-671616a1dbcb.png)

This is a mistake, the table is the same for all F3 and the CubeMX data confirms this. ST needs to get their shit together.

salkinium commented on tools/generator/dfg/stm32/stm.py:
```diff
@@ -89,6 +89,71 @@ def getGpioRemapForModuleConfig(module, config):
+        'tim6_up':   {'position': 13, 'clear': 1, 'set': 1},
+    },
+    'dma1ch4': {
+        'adc2':      {'position': 72, 'clear': 3, 'set': 3},
```
Oh gawd, this is a gem, see dma2ch1. Note that this writes both bit 72 and bit 73.

salkinium commented on tools/generator/dfg/stm32/stm.py:
```diff
@@ -89,6 +89,71 @@ def getGpioRemapForModuleConfig(module, config):
+
+
+    'dma2ch1': {
+        'adc2':     [{'position': 8, 'clear': 1, 'set': 0}, {'position': 73, 'clear': 1, 'set': 0}],
```
Right, so… ADC2 can be remapped entirely to DMA2 or DMA1 with bit 73:
<img width="411" alt="screen shot 2018-09-19 at 20 06 35" src="https://user-images.githubusercontent.com/163066/45772332-aa42bb00-bc47-11e8-9a6c-4e1477fb1393.png">
But this bit doesn't set the channel, nono! DMA 1 channel is selected with bit 72:
<img width="411" alt="screen shot 2018-09-19 at 20 06 35 copy" src="https://user-images.githubusercontent.com/163066/45772387-cd6d6a80-bc47-11e8-8723-78f6c2c97188.png">
And DMA2 channel is selected with bit 8, *which also remaps ADC4*.
<img width="607" alt="screen shot 2018-09-19 at 20 06 23" src="https://user-images.githubusercontent.com/163066/45772397-d65e3c00-bc47-11e8-8837-8e7dbce71505.png">

Good lord, ST, really?

salkinium commented on tools/generator/dfg/stm32/stm.py:
```diff
@@ -89,6 +89,142 @@ def getGpioRemapForModuleConfig(module, config):
+    },
+}
+
+stm32f0_dma_remap = \
```
@rleh @chris-durand These remaps are some ugly constraints, you're going to have to be smart about the channel scheduler with this.

salkinium commented:
I've added the remaps groups now in an automated fashion, please do a spot check for your devices, to see if there are any systematic errors in this. I've extracted the F0/F3 remap sets manually from the reference manuals, the may be faults in there.

salkinium committed:
[stm32] Update device files

salkinium commented:
I'll be merging this into modm-devices to prevent this work going stale and to allow implementing an (incomplete) API in modm already.
